nri_device_injector/nri_device_injector.go

```diff
@@ -0,0 +1,184 @@
+	Path     string `json:"path"`
+	Type     string `json:"type"`
+	Major    int64  `json:"major"`
+	Minor    int64  `json:"minor"`
+	FileMode uint32 `json:"file_mode"`
+	UID      uint32 `json:"uid"`
+	GID      uint32 `json:"gid"`
+}
+
+type plugin struct {
+	stub stub.Stub
+}
+
+func main() {
+	var (
+		opts []stub.Option
+		err  error
+	)
+
+	opts = append(opts, stub.WithPluginName(pluginName))
+	opts = append(opts, stub.WithPluginIdx(pluginIdx))
+
+	p := &plugin{}
+
+	if p.stub, err = stub.New(p, append(opts, stub.WithOnClose(p.onClose))...); err != nil {
+		log.Errorf("Failed to create plugin stub: %v", err)
+		os.Exit(1)
+	}
+
+	err = p.stub.Run(context.Background())
+	if err != nil {
+		log.Errorf("plugin exited with error %v", err)
+		os.Exit(1)
+	}
+}
+
+func (p *plugin) onClose() {
+	log.Info("NRI connection closed")
+}
+
+// CreateContainer handles CreateContainer requests relayed to the plugin by containerd NRI.
+// The plugin makes adjustment on containers with device injection annotations.
+func (p *plugin) CreateContainer(_ context.Context, pod *api.PodSandbox, container *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error) {
+	var (
+		ctrName string
+		devices []device
+		err     error
+	)
+
+	ctrName = container.Name
+	log.Infof("Started CreateContainer on container: %v, pod: %v, namespace: %v", ctrName, pod.Name, pod.Namespace)
+	if pod != nil {
+		ctrName = pod.Name + "/" + ctrName
+	}
+
+	devices, err = getDevices(container.Name, pod.Annotations)
+	if err != nil {
+		return nil, nil, err
+	}
+	adjust := &api.ContainerAdjustment{}
+
+	if len(devices) == 0 {
+		log.Infof("Container %s: no devices annotated...", ctrName)
+		return adjust, nil, nil
+	}
+	for _, d := range devices {
+		log.Infof("Container %s: annotated device %q...", ctrName, d.Path)
+		deviceNRI, err := d.toNRIDevice()
+		if err != nil {
+			return nil, nil, err
+		}
+		adjust.AddDevice(deviceNRI)
+		log.Infof("Container %s: injected device %q...", ctrName, d.Path)
+	}
+	return adjust, nil, nil
+}
+
+// getDevices returns parsed devices from pod annotations of device injections.
+func getDevices(ctrName string, podAnnotations map[string]string) ([]device, error) {
+	var (
+		deviceKey string = ctrDeviceKeyPrefix + ctrName
+
+		annotation []byte
+		devices    []device
+	)
+
+	if value, ok := podAnnotations[deviceKey]; ok {
+		annotation = []byte(value)
+	}
+	if annotation == nil {
+		return nil, nil
+	}
+
+	if err := yaml.Unmarshal(annotation, &devices); err != nil {
+		return nil, fmt.Errorf("invalid device annotation %q: %w", deviceKey, err)
+	}
+	return devices, nil
+}
+
+// toNRIDevice retrieves device's major, minor and type from its path, and returns a NRI device
+func (d *device) toNRIDevice() (*api.LinuxDevice, error) {
+	var (
+		stat unix.Stat_t
+	)
+	if err := unix.Lstat(d.Path, &stat); err != nil {
+		return nil, err
+	}
+
+	var (
+		devNumber = uint64(stat.Rdev) //nolint:nolintlint,unconvert // the type is 32bit on mips.
```

Comment:
I don't think we're running a linter here? unconvert or nolintlint?

(Adding a linter is a good idea, but if we don't have one let's remove the directive for now and add it back when we need it.)

Reply:
it seems this repo hs a linter https://github.com/GoogleCloudPlatform/container-engine-accelerators/actions/runs/7295051489/job/19880874295?pr=345

Reply:
Yes, https://github.com/GoogleCloudPlatform/container-engine-accelerators/blob/2ba1c84deee6cb50ede81e132ed10b9ac6d7a748/Makefile#L27-L35, but that wouldn't include the unconvert or nolintlint linters.

Reply:
Thanks Sam, I've removed the nolint comment